vm/ubpf_verifier.c

```diff
@@ -44,6 +37,57 @@ int isjmp(struct ebpf_inst inst)
     return false;
 }
 
+int uses_src(struct ebpf_inst inst)
+{
+    char cls = inst.opcode & EBPF_CLS_MASK;
+    if (inst.opcode == EBPF_OP_EXIT) {
+        return true;
+    } else if ((cls == EBPF_CLS_STX) ||
+        (cls == EBPF_CLS_LDX)) {
+        return true;
+    } else if (((cls == EBPF_CLS_ALU) ||
+         (cls == EBPF_CLS_ALU64) ||
+         (cls == EBPF_CLS_JMP)) &&
+         (inst.opcode & EBPF_SRC_REG)) {
+        // Non-trivial exceptions
+        if ((inst.opcode == EBPF_OP_NEG) ||
+            (inst.opcode == EBPF_OP_NEG64) ||
+            (inst.opcode == EBPF_OP_LE) ||
+            (inst.opcode == EBPF_OP_BE) ||
+            (inst.opcode == EBPF_OP_LDDW) ||
+            (inst.opcode == EBPF_OP_JA) ||
+            (inst.opcode == EBPF_OP_CALL)) {
+            return false;
+        } else {
+            return true;
+        }
+    } else {
+        return false;
```

Comment:
I would prefer to list explicitly all the `return false` cases, and default to `return true`. We're less likely to have false negatives that way.